### PR TITLE
Add Priority to Kubernetes API

### PR DESCRIPTION
**Pull request by bsalamat**
**What this PR does / why we need it**: This is the first in a series of PRs to add priority to Kubernetes API. Subsequent PRs will add priority name resolution to admission controller.

**Release note**:

```release-note
Add PriorityClassName and Priority fields to PodSpec.
```

### CODE REVIEW

**Comment by bsalamat** [comment]
ref/ #22212

**Comment by bsalamat** [comment]
@kubernetes/api-reviewers
@kubernetes/sig-scheduling-pr-reviews

**Comment by timothysc** [comment]
Needs a release note + xref: the proposal.  

@kubernetes/sig-api-machinery-feature-requests folks will need to sign off on this one.

**Comment by bsalamat** [comment]
ref/ kubernetes/community#604

File: pkg/api/v1/types.go
Comment (davidopp) [inline pkg/api/v1/types.go]: Add a TODO to update this with the name of the admission controller once you write it.
BTW do you think you can do the admission controller for 1.7? It shouldn't be too hard if you mostly copy one of the existing admission controllers. I don't think any of the existing admission controllers are configured by a ConfigMap, though.

File: pkg/api/v1/types.go
Comment (davidopp) [inline pkg/api/v1/types.go]: Same thing here, update with actual name of the admission controller once you write it.

File: pkg/api/validation/validation.go
Comment (davidopp) [inline pkg/api/validation/validation.go]: A couple of comments about validation

(1) You should also do validation on Priority (not just PriorityName) even though it is set by an admission controller (user might write a buggy custom admission controller). BTW do we want to allow negative priorities?

(2) I think you need to modify ValidatePodUpdate() to make it possible to update PriorityName (and Priority)

File: pkg/api/types.go
Comment (k82cn) [inline pkg/api/types.go]: zero is the lowest?

**Comment by liggitt** [comment]
who is allowed to set and update a pod's priority?

**Comment by liggitt** [comment]
Is this proposing that the same priorityName would resolve to different numerical priorities for different users? e.g. I (as an end-user) submit a pod with "priorityName: system" and get priority of 0, and you (as a superuser) submit the same pod and get a priority of 1000?

Or would I be prevented from creating a pod with `priorityName: system` at all?

**Comment by timothysc** [comment]
>Is this proposing that the same priorityName would resolve to different numerical priorities for different users? e.g. I (as an end-user) submit a pod with "priorityName: system" and get priority of 0, and you (as a superuser) submit the same pod and get a priority of 1000? 

Not yet, but I did have that intention in the long run.  

> Or would I be prevented from creating a pod with priorityName: system at all? 

I hope we would lock down the highest priority to admin only.

**Comment by liggitt** [comment]
>> Or would I be prevented from creating a pod with priorityName: system at all?
>
> I hope we would lock down the highest priority to admin only.

Lock down by rejecting the pod, or by translating the `system` priority name to a lower numeric priority? I want to make sure we have an idea how we would do this (and how we would decide whether pods created by controllers should be allowed to have system priority). That seems like an important flow to nail down before making API changes.

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: No, priority could potentially be negative. Zero is default.

**Comment by bsalamat** [comment]
@liggitt Priority is meant to be global, not per user or namespace. Users can specify PriorityName is their podspec. `system` is a special priority for system pods. I agree that we should limit it to admin so that only admin can set it. Pods by other users who put `system` in their pod spec will be rejected. I will add this to the design proposal.
This PR only adds the fields with almost no enforcement logic.

File: pkg/api/v1/types.go
Comment (bsalamat) [inline pkg/api/v1/types.go]: Let me call it "Priority" admission controller for now. If we decided that it is not a good name, I will change it.
And, yes, I would like to do admission controller for 1.7 as well if we reach a consensus on how priority should be defined and work.

File: pkg/api/validation/validation.go
Comment (bsalamat) [inline pkg/api/validation/validation.go]: 1. Actually any valid integer is a valid priority, even negative numbers. I am not sure if we can do any particular validation for "Priority".

2. I was not so sure whether we want to allow updating priority or not. I am going to modify ValidatePodUpdate() so that it allows change of priority with the assumption that all updates go through admission checks.

File: api/swagger-spec/v1.json
Comment (gyliu513) [inline api/swagger-spec/v1.json]: Juts nit: Those should be the generated files? As you are going to have a new patch based on the new proposal, so I think that you will update here also.

Comment (bsalamat) [inline api/swagger-spec/v1.json]: I will update all the generated files eventually. For some reason, it is not that easy to update them. If I run hack/update-all.sh it does not work and corrupts my files. I have to run individual update scripts in a not-so-well-known order to make it work.

**Comment by bsalamat** [comment]
Given the discussion on ubernetes/community#604, this PR is in a good state.
@davidopp @timothysc could you take another look?

File: api/swagger-spec/v1.json
Comment (gyliu513) [inline api/swagger-spec/v1.json]: Is it a bug that the `hack/update-all.sh` does not work and corrupts your file?

Comment (bsalamat) [inline api/swagger-spec/v1.json]: No, turned out to be a problem with my Go SDK. I had updated my Go installation, but hadn't cleaned my repository after the update. Running `make clean` and rebuilding the project fixed the issue.

File: pkg/api/types.go
Comment (k82cn) [inline pkg/api/types.go]: Will admission controller set it to lowest priority? The concern is that: if not specified, there maybe an undefined priority in system, e.g. zero is not a defined priority.

File: pkg/api/v1/types.go
Comment (k82cn) [inline pkg/api/v1/types.go]: Can we also add definition of `PriorityClass`?

Comment (bsalamat) [inline pkg/api/v1/types.go]: Yes, I wanted to add it in a separate PR though.

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: I updated the proposal a few minutes ago and added more details about the default name resolution. We probably go with resolving unspecified PriorityClassName to zero in the initial version, but in a follow up change, we will add the support to annotate one of the PriorityClasses with a "default" annotation. Once such a priority class exists, the system resolves unspecified PriorityClassName to the value of the default class. If there is no such a class, we will still resolve to zero. Zero is not the lowest possible priority in the system given that any valid integer is a valid priority.

**Comment by timothysc** [comment]
@bsalamat please adjust the title and commit from [WIP]

**Review by timothysc — COMMENTED** [review]
It doesn't look like we are doing anything with the fields yet.  What is the intended deliverable for 1.7?

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: Tim Hockin, Jordan and Wojtek raised the concern that setting pod spec should not directly affect status.
This is from Jordan:
> Currently, setting the spec explicitly does not affect the status (and vice versa) (the create/update strategy for all objects with spec preserves the existing status as-is, and the status strategy preserves the existing spec as-is). This changes that behavior for pods, and means that modifying spec can conflict with status updates, and anything that can modify status can change "effective" pod priority.

This is from thockin:

> There's a rule of thumb that everything in status can be recreated from observation. This would not be the case here, as (I assume) priority name->num mapping could change after a pod was created.
To that end, it SHOULD be part of spec, but then you have to answer what happens if a user sets it directly. I would guess you want to disallow that.

**Comment by bsalamat** [comment]
> It doesn't look like we are doing anything with the fields yet. What is the intended deliverable for 1.7?

I am adding a couple more PRs for 1.7. One PR to add PriorityClass and another one is an admission controller to set Priority. Kubelet team wants to incorporate priority in their out of resource preemption logic in 1.7.

**Comment by bsalamat** [comment]
@davidopp @timothysc 
My subsequent PR to add PriorityClass API object is also ready: https://github.com/bsalamat/kubernetes/pull/1

I am waiting for this PR to be merged and then I will send that one.

**Comment by smarterclayton** [comment]
Is the preemption for QoS vs priority discussion settled?  From the unmerged proposal, I don't see agreement.  Who on sig-node will be owning merging that?

Without an actual use in 1.7, I don't want to add API fields.

File: pkg/api/types.go
Comment (smarterclayton) [inline pkg/api/types.go]: We use `CamelCase` for all constants, if this is a "opaque constant field".  If this is a reference to a priority class API object, the field name should be `PriorityClassName`.

File: pkg/api/v1/types.go
Comment (smarterclayton) [inline pkg/api/v1/types.go]: I don't think the admission controller name should be in the API - that has nothing to do with the end user's perception of the field.  A custom admission controller can handle this as well as anything else.

File: pkg/api/validation/validation.go
Comment (smarterclayton) [inline pkg/api/validation/validation.go]: Please don't allow priorityName to be updated unless you are absolutely sure you are going to use it in 1.7.  If you aren't, leave it create only.  We can never tighten validation.

File: pkg/api/types.go
Comment (k82cn) [inline pkg/api/types.go]: @bsalamat , "default" annotation makes sense to me :).

Comment (bsalamat) [inline pkg/api/types.go]: @smarterclayton This is a reference to a priority class object. The field is already named `PriorityClassName`.

@k82cn that's the plan.

**Comment by bsalamat** [comment]
@smarterclayton kubelet preemption is not finalized. @dashpole is the owner of out of resource preemption on the node. We still might be able to finalize and have it in 1.7 in the next several days, but I am not optimistic.

That said, we need to add the fields (but not necessarily in 1.7) so that we can add subsequent PRs for adding the PriorityClass API object and the admission controller.

**Comment by vishh** [comment]
@smarterclayton regarding interactions with QoS:
@dashpole and I are working with @bsalamat on how the Kubelet would handle priority for evictions.
Current plan is to not make any changes to kubelet evictions in v1.7. @dashpole will post a proposal on how evictions will be updated to take priority into account.

tl;dr; is that kubelet will lower priority on pods proportional to it's usage above requests and choose pods to be evicted based on priorities.

Why do you think introducing a notion of priority has to wait for preemptions design to settle?

**Comment by davidopp** [comment]
>  Kubelet team wants to incorporate priority in their out of resource preemption logic 

Nit: I think it's clearer if we call what happens when there is out-of-resource "eviction logic" rather than "preemption logic."

File: pkg/api/validation/validation.go
Comment (thockin) [inline pkg/api/validation/validation.go]: This allows changes from "foo" to "" - is that intended?  what is the semantic?

File: pkg/api/validation/validation.go
Comment (thockin) [inline pkg/api/validation/validation.go]: This warrants a comment - a subsequent reader would be very confused.

File: pkg/api/validation/validation.go
Comment (thockin) [inline pkg/api/validation/validation.go]: You already asserted that class is the same and name is nil - you don't need these?

File: pkg/api/types.go
Comment (thockin) [inline pkg/api/types.go]: Can we make the "special" name be something that is out of bounds for real instances?  Maybe "SYSTEM" or "kubernetes.io/system" or "auto:system" or something?  I don't want to debug clusters that have an instance of PriorityClass (where is that part defined, btw - I think it should be this same PR) named "system".

File: pkg/api/types.go
Comment (thockin) [inline pkg/api/types.go]: This (or the versioned copy) should also say how to set the value, since it is validated to be `== nil`

File: pkg/api/validation/validation.go
Comment (bsalamat) [inline pkg/api/validation/validation.go]: Yes, this is allowed. An empty priority class name is resolved to the "default" priority. One of the priority classes can be annotated as default and its value will be used as default priority.

File: pkg/api/validation/validation.go
Comment (bsalamat) [inline pkg/api/validation/validation.go]: I guess you are referring to the code at line 2576-2580. Those lines verify that if the class name has changed and the new name is not empty, the new name is a valid name.

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: We actually wanted to use "system" for critical system pods. I am fine with the idea of the making this a special keyword that pods cannot use, but then we will need to ship Kubernetes with at least one more predefined priority for system pods.
The PR to add PriorityClass is branched off of this branch and currently lives in my repo until this one is merged: https://github.com/bsalamat/kubernetes#1

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: The value should not be set. It is a pointer and if not provided, will be nil.

**Comment by bsalamat** [comment]
I guess I am not going to rebase again before getting lgtm. The reason for needing a rebase is a large number of auto-generated files (we will hopefully remove them from the repo in the near future).

File: pkg/api/validation/validation.go
Comment (thockin) [inline pkg/api/validation/validation.go]: You are disallowing "foo" -> "bar", but "foo" -> "" is OK?  why?  If this is right, comment why, but explain it first?

File: pkg/api/validation/validation.go
Comment (thockin) [inline pkg/api/validation/validation.go]: I have an open question there about why changing to "" is OK.

File: pkg/api/types.go
Comment (thockin) [inline pkg/api/types.go]: I don't understand.  You want system pods to use the built-in, special priority.  I'm just saying that the name of that special prio should be something that can't be created by a user (hence the options of names that are invalid resource names)

File: pkg/api/types.go
Comment (thockin) [inline pkg/api/types.go]: As someone who reads the API, I'm seeing that this is an optional field, but validation won't allow any value to be set.  How does the admission controller set the value?

File: pkg/api/validation/validation.go
Comment (bsalamat) [inline pkg/api/validation/validation.go]: No, I am not disallowing "foo" -> "bar". That condition just checks that if there is a change, then it validates the new name. It doesn't prevent the change of priority class name.

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: Agreed. We should find a better naming convention for default priority class name(s). The choice of names won't affect this PR (except for the comment here) though.

Comment (thockin) [inline pkg/api/types.go]: agree.  Can we say "SYSTEM" here so it's at least non-conflicting by default ?

File: pkg/api/validation/validation.go
Comment (thockin) [inline pkg/api/validation/validation.go]: Ahh, I mis-read.  Why are you allowing any update to priority name?

Comment (bsalamat) [inline pkg/api/validation/validation.go]: To allow a user to change the priority of a running job.

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: Sure, I will do so.

**Comment by thockin** [comment]
Do we NEED to allow that?  Almost nothing else about pod is mutable (yet).
If so, maybe comment that in the API comments, including which changes are
allowed and semantics.

**Comment by bsalamat** [comment]
> Do we NEED to allow that?  Almost nothing else about pod is mutable (yet).
If so, maybe comment that in the API comments, including which changes are
allowed and semantics.

I think it will be a useful feature, but a cluster will certainly work without it. I think @davidopp wanted to have this feature. What do you think David about removing it?

**Comment by thockin** [comment]
Do we expect a change to this field to be actuated in real-time by kubelet
(e.g. into oom_adjust and other mechanisms)?  That seems onerous for v1 ?

**Comment by bsalamat** [comment]
@thockin I agree. It may be too much to expect from Kubelet to apply the changes in real-time in v1. I'm fine with removing it, if David has no objections.

**Comment by smarterclayton** [comment]
I would prefer leaving it as immutable for now as well.

File: pkg/api/v1/types.go
Comment (ravisantoshgudimetla) [inline pkg/api/v1/types.go]: Shouldn't these 2 lines of comments be used for priority instead of PriorityClassName?

Comment (bsalamat) [inline pkg/api/v1/types.go]: No, they are placed correctly, but if they are confusing and you have a better suggestion, I will gladly fix them. `Priority` should never be set by users.

Comment (ravisantoshgudimetla) [inline pkg/api/v1/types.go]: Thanks for the explanation @bsalamat . The way I was thinking about it is - "If this is not specified, the `'Priority'` field of pod will be set to zero"

Comment (bsalamat) [inline pkg/api/v1/types.go]: @ravisantoshgudimetla If this filed is not set, the `priority` field of the pod will be set to zero (or default priority if exists).

**Comment by davidopp** [comment]
@bsalamat Can you summarize the remaining issues (if any) blocking this PR from being merged?

**Comment by bsalamat** [comment]
>  Can you summarize the remaining issues (if any) blocking this PR from being merged?

There is no remaining issue. The last thing I was asked to do was to remove the possibility of updating `PriorityClassName`. I addressed that a while ago.

@timothysc Does this PR look good to you?

**Comment by timothysc** [comment]
@bsalamat Needs rebase + remove WIP from title

**Comment by ravisantoshgudimetla** [comment]
@bsalamat The design proposal against this PR is still open, shouldn't we merge that before we proceed with this PR?

**Comment by bsalamat** [comment]
@ravisantoshgudimetla I believe kubernetes/community/pull/604 is also ready to merge as there is no more unaddressed comments.

**Comment by ravisantoshgudimetla** [comment]
@bsalamat Cool, I was concerned about the order.

**Comment by bsalamat** [comment]
@timothysc Mostly done, but it is annoying that by the time I finish updating the autogenerated files, those files are out-of-sync again. I will try again to update autogenerated files.

**Comment by bsalamat** [comment]
@k8s-bot pull-kubernetes-node-e2e test this

File: pkg/api/types.go
Comment (timothysc) [inline pkg/api/types.go]: We should make certain admission control always overrides this value.  I'm still concerned of users blindly mucking this value.

File: pkg/api/validation/validation.go
Comment (liggitt) [inline pkg/api/validation/validation.go]: both fields should be prevented from being populated if the priority feature gate is not on... we don't want latent data allowed in, then acted on in a later release when the feature is enabled

File: pkg/api/types.go
Comment (bsalamat) [inline pkg/api/types.go]: Admission controller rejects a pod if its `Priority` is set by user. I have already implemented it, but since those are branched from this PR, I cannot send them for review yet.

File: pkg/api/validation/validation.go
Comment (bsalamat) [inline pkg/api/validation/validation.go]: Added the feature gate for PriorityClassName. Admission controller (a subsequent PR) will prevent users from setting `Priority` regardless. `Priority` is set only be the admission controller.

Comment (liggitt) [inline pkg/api/validation/validation.go]: The rest handler cannot depend on the admission plug-in being present, and by the time the object reaches validation and rest handling, there is no way to know whether the field was set by a user or an admission plug-in

Comment (bsalamat) [inline pkg/api/validation/validation.go]: Makes sense. I added a check for preventing `Priority` from being set if the feature is not enabled.

**Comment by timothysc** [comment]
I'm ok with this one, b/c it sounds like there are follow on PRs.  Could you open a checklist, or add a checklist to the feature issue, so we can check across the release.

**Comment by bsalamat** [comment]
ref/ #47604

**Comment by timothysc** [comment]
@k82cn , @liggitt , @davidopp - I'm good with what is here now and the checklist for 1.8, pending on your lgtm.

File: pkg/api/validation/validation_test.go
Comment (liggitt) [inline pkg/api/validation/validation_test.go]: not sure how setting globals will affect tests that run in parallel, but at least obtain and defer restoring the original value

File: staging/src/k8s.io/client-go/pkg/api/types.go
Comment (liggitt) [inline staging/src/k8s.io/client-go/pkg/api/types.go]: Documenting this in terms of an admission plugin reads really strangely. The API field should not require an admission plugin to make sense. If the admission plugin and the behavior described is required, then it probably shoudn't be an admission plugin, and should be part of the rest handler.

File: pkg/api/validation/validation_test.go
Comment (bsalamat) [inline pkg/api/validation/validation_test.go]: Looks like it is a common practice in our tests and in that file. Anyway, I added the code to obtain and restore the value.

File: staging/src/k8s.io/client-go/pkg/api/types.go
Comment (bsalamat) [inline staging/src/k8s.io/client-go/pkg/api/types.go]: That's a good point. I changed the comment and tried to explain how it works.

File: pkg/api/types.go
Comment (gyliu513) [inline pkg/api/types.go]: s/will will/will

Comment (gyliu513) [inline pkg/api/types.go]: `will be default or zero if there is no default`.

Are we going to support `default` now? I found that in the design document, it is `will`, just want to confirm if the `default` will be available in the first implementation? If not, how about add some words such as `will` to identify it is future work?

Comment (bsalamat) [inline pkg/api/types.go]: The plan to have the default in the first version. We will definitely have the ability to define a "global" default priority. In the future versions, we may add per namespace default as well.

Comment (bsalamat) [inline pkg/api/types.go]: Done. Thanks!

Comment (gyliu513) [inline pkg/api/types.go]: Got it, then how about update the sentence `default or zero if there is no default` a bit?

Such as when using default one and when setting to zero? Using `default or zero if there is no default` may make people confused, as they may not clear when using default and when using zero?

File: pkg/api/v1/types.go
Comment (k82cn) [inline pkg/api/v1/types.go]: different comments with `api/types.go`, should be `by creating a PriorityClass`.

File: pkg/api/validation/validation.go
Comment (k82cn) [inline pkg/api/validation/validation.go]: personally, I'd like the following if/else:

```
if utilfeature.DefaultFeatureGate.Enabled(features.PodPriority) {
  ...
} else {
  ...
}
```

**Review by k82cn — APPROVED** [review]
Overall, it's LGTM :); just some comments on cleanup.

**Comment by k82cn** [comment]
btw, please run `./hack/update-staging-client-go.sh` for `client-go`.

File: pkg/api/validation/validation.go
Comment (bsalamat) [inline pkg/api/validation/validation.go]: in my version of the code, the feature gate is not checked when priority class is not specified. Also, the condition "if len(spec.PriorityClassName) > 0" should be repeated in both if and else blocks of your version.

Comment (k82cn) [inline pkg/api/validation/validation.go]: yes; my version is for sometimes removing this feature gate (just remove the `else` block is OK); both are ok to me :).

**Comment by gyliu513** [comment]
@bsalamat how about run `./hack/update-all.sh -a`? This can help generate code and update client-go, also have some CLI interaction to enable us to submit different commits.

**Comment by bsalamat** [comment]
@gyliu513 I have already. Did you see files which are not updated?

**Comment by gyliu513** [comment]
@bsalamat I found that `dd priority to Kubernetes API` also include some generated code, and also seems your patch do not updated the code for `client-go`.  This seems strange to me.

But if you use `./hack/update-all.sh -a`, you can first get the generated code for a commit and then another `client-go` update for another commit.

**Comment by bsalamat** [comment]
@smarterclayton Could you please take another look?

**Pull request merged**
